### ISSUE TITLE
Weird error message in torch.split_size_or_sections

### ISSUE THREAD

**Issue by Diego999** [pytorch/pytorch#5212]
Hi everyone,

I wanted to use the new function torch.split with a list of integer.

I have a input vector of dim X (let's say 100) and a list of integer where the sum is less or equal (Y <= X). If Y <= X, an error is raised saying "Sum of split sizes exceeds tensor dim", which is not the case.

`
def split(tensor, split_size_or_sections, dim=0):

    if dim < 0:
        dim += tensor.dim()
    dim_size = tensor.size(dim)
    if isinstance(split_size_or_sections, int):
        split_size = split_size_or_sections
        num_splits = (dim_size + split_size - 1) // split_size
        last_split_size = split_size - (split_size * num_splits - dim_size)

        def get_split_size(i):
            return split_size if i < num_splits - 1 else last_split_size
        return tuple(tensor.narrow(int(dim), int(i * split_size), int(get_split_size(i))) for i
                     in range(0, num_splits))

    else:
        if dim_size != sum(split_size_or_sections):
            raise ValueError("Sum of split sizes exceeds tensor dim")
        split_indices = [0] + split_size_or_sections
        split_indices = torch.cumsum(torch.Tensor(split_indices), dim=0)

        return tuple(
            tensor.narrow(int(dim), int(start), int(length))
            for start, length in zip(split_indices, split_size_or_sections))
`

In this case, either we should have if dim_size <= sum(split_size_or_sections), either we should split the tensor is chuncks where the sum is lower than dim_size.

OS: Mac OS X
PyTorch version: From source (https://github.com/pytorch/pytorch/commit/2b2d56d8460d335daf5aa79774442a111d424f90)
How you installed PyTorch (conda, pip, source): pip
Python version: 3.6

**Comment by ptrblck** [comment]
Thanks for reporting this issue @Diego999.
I haven't thought of a use case with splits < dim.

What do you think the appropriate solution would be?
Discarding the additional slices or return a separate split?

E.g.:
``
x = torch.randn(10, 20)
torch.split(x, [5, 2], 0)
``
In your use case, should it return two Tensors of size `[5, 20]` and `[2, 20]` or another with `[3, 20]`?

**Comment by Diego999** [comment]
Thanks for your answer @ptrblck 

I was expecting to have 3 tensors w.r.t your example: [5,20], [2,20] and another with the rest (here [3, 20]).

What do you think ?

**Comment by ptrblck** [comment]
Sounds reasonable to me.
This behavior would be similar to what `torch.split` returns using an integer for the split size.

``
x = torch.randn(10, 20)
torch.split(x, 4, 0)
``
returns 3 splits of shape
`[4, 20]`, `[4, 20]`, `[2, 20]`.

**Comment by Diego999** [comment]
Yes, exactly ! The only difference is now the function should infer the rest if the sum != size

**Comment by Diego999** [comment]
Would you like I modify the code and do a pull request ?

**Comment by ptrblck** [comment]
@Diego999 I already have the code on my machine, but was talking to @apaszke about the right syntax.
Summary:

One issue is that numpy's `.split` function behaves differently than the current Pytorch implementation. I'm no sure how to deal with this, since I like the PyTorch behavior, but maybe the core devs would like to make the API as numpy-like as possible.

* `np.split` takes an integer or a list. The integer defines the number of `sections`, while the list defines `split_indices`. 
* `torch.split` currently also takes an integer or a list. The integer defines the `split_size`, while the list defines `split_sizes` for each section.

**numpy integer example**:
```
a = np.random.rand(10, 20)
tmp0 = np.split(a, indices_or_sections=5, axis=0) # split into 5 sections
for t in tmp0:
    print(t.shape)
```
This will return 5 sections, each of a size of `[2, 20]`.

```
np.split(a, indices_or_sections=7, axis=0) # error, since no equal division
```
Won't work, since `axis=0` is not divisible by 7.

**PyTorch integer example**:
```
x = torch.randn(10, 20)
tmp2 = torch.split(x, split_size_or_sections=4, dim=0) # use size 4
for t in tmp2:
    print(t.shape) # last split might be smaller
```
Will return 3 splits with dims: `[4, 20], [4, 20], [2, 20]`. `dim=0` does not have to be divisible by `split_size_or_sections`. In my opinion it's more user-friendly than numpy's approach.

**numpy list example**:
```
tmp1 = np.split(a, [5, 7], 0) # use indices ([:5], [5:7], [7:])
for t in tmp1:
print(t.shape)
```
Will split `a` into 3 sections using the list entries as indices: `[:5], [5:7], [7:]`.

**PyTorch list example**:
```
tmp3 = torch.split(x, split_size_or_sections=[5, 2, 3], dim=0)
for t in tmp3:
    print(t.shape)
```
Will return 3 splits with the defined dimensions: `[5, 20], [2, 20], [3, 20]`. Currently the sum of the list entries has to be equal to the specified dimension size of the `Tensor`.

As @Diego999 suggested we could modify the code easily to be less strict about the sizes:
```
torch.split(x, split_size_or_sections=[5, 4], dim=0) # currently error, since 5+4 != dim(0)
# Could return Tensors of size [5, 20], [4, 20] and [1, 20]
```

What do the others think about this issue?
Are we good to go with relaxing the size constraints of `torch.split`?
How should we deal with numpy's approach of splitting using indices instead of sizes?

- Add another split function / refactor the current `torch.split`

- Specify a flag to use the list as sizes or indices, e.g. `use_as_index=False`

- Just keep the PyTorch approach and hope numpy users won't be confused

[Here](https://gist.github.com/ptrblck/a04b4ba8d641267746c2e85b3da6c899) is a complete gist for reproduction.

**Comment by Diego999** [comment]
@ptrblck Thank you for this update !

**Comment by ngimel** [comment]
The original issue with incorrect error message is solved (current error message is `plit_with_sizes expects split_sizes to sum exactly to 4 (input tensor's size at dimension 0), but got split_sizes=[2, 1]`), adding `numpy` label for the different behaviors of torch and numpy `split`.

**Comment by rgommers** [comment]
> Add another split function / refactor the current `torch.split`

This is very rarely used functionality (see, e.g., [usage data for numpy](https://raw.githubusercontent.com/data-apis/python-record-api/master/data/typing/numpy.py)), so I'd be -1 on a second function.

> Specify a flag to use the list as sizes or indices, e.g. `use_as_index=False`

That seems reasonable.

> Just keep the PyTorch approach and hope numpy users won't be confused

I don't think many people will be confused, but the keyword seems easy to add and sensible - so with the goal of improving numpy compatibility where possible, why not add that keyword.
